modules/apps/object/object-web/src/main/java/com/liferay/object/web/internal/object/definitions/portlet/action/GetObjectRelationshipEdgeCandidatesMVCResourceCommand.java

```diff
@@ -0,0 +1,213 @@
+
+			if (!_meetsTreeMaxHeight(depth, parentObjectDefinition, tree)) {
+				continue;
+			}
+
+			List<Edge> edges = tree.getAncestorsPath(
+				parentObjectDefinition.getObjectDefinitionId());
+
+			JSONArray ancestorsJSONArray = jsonFactory.createJSONArray();
+
+			for (Edge edge : edges) {
+				ancestorsJSONArray.put(
+					JSONUtil.put(
+						"label",
+						() -> {
+							ObjectRelationship edgeObjectRelationship =
+								_objectRelationshipLocalService.
+									getObjectRelationship(
+										edge.getObjectRelationshipId());
+
+							ObjectDefinition objectDefinition =
+								_objectDefinitionLocalService.
+									getObjectDefinition(
+										edgeObjectRelationship.
+											getObjectDefinitionId1());
+
+							return StringUtil.appendParentheticalSuffix(
+								objectDefinition.getLabel(locale),
+								edgeObjectRelationship.getLabel(locale));
+						}
+					).put(
+						"objectRelationshipId", edge.getObjectRelationshipId()
+					));
+			}
+
+			objectRelationshipsJSONArray.put(
+				JSONUtil.put(
+					"ancestors", ancestorsJSONArray
+				).put(
+					"label",
+					StringUtil.appendParentheticalSuffix(
+						parentObjectDefinition.getLabel(locale),
+						objectRelationship.getLabel(locale))
+				).put(
+					"objectRelationshipId",
+					objectRelationship.getObjectRelationshipId()
+				).put(
+					"root",
+					parentObjectDefinition.getObjectDefinitionId() ==
+						parentObjectDefinition.getRootObjectDefinitionId()
+				));
+		}
+
+		JSONPortletResponseUtil.writeJSON(
+			resourceRequest, resourceResponse, objectRelationshipsJSONArray);
+	}
+
+	@Reference
+	protected JSONFactory jsonFactory;
+
+	private long _getObjectDefinitionId(ResourceRequest resourceRequest)
+		throws Exception {
+
+		long objectRelationshipId = ParamUtil.getLong(
+			resourceRequest, "objectRelationshipId");
+
+		if (objectRelationshipId != 0) {
+			ObjectRelationship objectRelationship =
+				_objectRelationshipLocalService.getObjectRelationship(
+					ParamUtil.getLong(resourceRequest, "objectRelationshipId"));
+
+			return objectRelationship.getObjectDefinitionId1();
+		}
+
+		return ParamUtil.getLong(resourceRequest, "objectDefinitionId");
+	}
+
+	private boolean _meetsEdgeCriteria(ObjectRelationship objectRelationship)
+		throws Exception {
+
+		ObjectDefinition parentObjectDefinition =
+			_objectDefinitionLocalService.getObjectDefinition(
+				objectRelationship.getObjectDefinitionId1());
+
+		if (objectRelationship.isSelf() ||
+			!Objects.equals(
+				ObjectRelationshipConstants.TYPE_ONE_TO_MANY,
+				objectRelationship.getType()) ||
+			parentObjectDefinition.isUnmodifiableSystemObject()) {
+
+			return false;
+		}
+
+		return true;
+	}
+
+	private boolean _meetsTreeMaxHeight(
+		int depth, ObjectDefinition objectDefinition, Tree tree) {
+
+		Node node = tree.getNode(objectDefinition.getObjectDefinitionId());
+
+		if ((node.getDepth() + depth) > 4) {
+			return false;
+		}
+
+		return true;
+	}
```

Comment:
@MarinhoFeliphe Put this method inside Tree implementation. Something like isMaxPathNodeHeight. Not sure about this one, what do you think?

Reply:
@gabrielwas I didn't changed a lot, I just created a constant to avoid this magic number

[changes](https://github.com/liferay-objects/liferay-portal/pull/2431/commits/691426bc0e8925019a234e4f0a9d3eabb8e7c21d)